src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -3274,21 +3278,274 @@ def _if_stmt_handler(self, node, parent):
+                     f"\"{guard_type_repr[idx].lower()}\"\n")
+            code += (f"    {pointer_name} => {selector}\n")
+        code += "end select\n"
+        code += "end program\n"
+        parser = ParserFactory().create(std="f2008")
+        reader = FortranStringReader(code)
+        fp2_program = parser(reader)
+        # Only use the select type clause within the fparser2 program
+        fp2_select_type = fp2_program.children[0].children[1]
+        code_block = CodeBlock(
+            [fp2_select_type], CodeBlock.Structure.STATEMENT, parent=parent)
+        # Handlers assume a single node is returned and in this
+        # implementation we create an assignment, a CodeBlock and an
+        # if statement (see later). Therefore we add the assignment
+        # and codeblock to the parent here and return the if statement
+        # in the usual way.
+        parent.addchild(Assignment.create(
+            Reference(type_string_symbol), Literal("", CHARACTER_TYPE)))
+        parent.addchild(code_block)
+
+        # Step4: Create the (potentially nested) if statement that
+        # replicates the functionality of the select type options (as
+        # select type is not supported directly in the PSyIR).
+        outer_ifblock = None
+        ifblock = None
+        currentparent = parent
+        for idx in range(select_idx+1):
+            if idx == default_idx:
+                continue
+            if ifblock:
+                # We already have an if so this is an else if
+                elsebody = Schedule(parent=currentparent)
+                currentparent.addchild(elsebody)
+                annotation = "was_type_is"
+                if clause_type[idx] == "CLASS IS":
+                    annotation = "was_class_is"
+                ifblock = IfBlock(annotations=[annotation])
+                elsebody.addchild(ifblock)
+            else:
+                annotation = "was_type_is"
+                if clause_type[idx] == "CLASS IS":
+                    annotation = "was_class_is"
+                ifblock = IfBlock(parent=currentparent,
+                                  annotations=[annotation])
+                outer_ifblock = ifblock
+
+            # Create an if hierarchy that uses the string (stored in
+            # type_string_symbol) set in the previous select type
+            # codeblock to choose the appropriate content in the
+            # original select type clauses.
+            clause = BinaryOperation.create(
+                BinaryOperation.Operator.EQ, Reference(type_string_symbol),
+                Literal(guard_type_repr[idx].lower(), CHARACTER_TYPE))
+
+            ifblock.addchild(clause)
+            # Add If_body
+            ifbody = Schedule(parent=ifblock)
+            self.process_nodes(parent=ifbody, nodes=stmts[idx])
+            # Replace references to the type selector variable with
+            # references to the appropriate pointer.
+            for reference in ifbody.walk(Reference):
```

Comment:
You'll need to check that we didn't end up with a CodeBlock when processing the body of the clause.

Reply:
Hmm. I'll need advice on this. Presumably I'll have to roll back to a NotImplementedError in this case as all pointers will need to be found within the code body (unless I look inside codeblocks which I'm not sure will be robust). This would need me to check at the beginning before any changes to the tree are made as in-place. Do you agree?

Reply:
I've done this. I create a fake parent and then only switch it for the real one at the end of the processing. Also added a test.